### ISSUE TITLE
[Question]: suggestion for See et al. copy mechanism implementation

### ISSUE THREAD

**Issue by aleSuglia** [allenai/allennlp#1727]
Hi there,

I was wondering if you have some recommendation in terms of how to implement a copy mechanism for a seq2seq model following AllenNLP best practices. I was thinking to modify the DataIterator in order to return, for each batch, an extended vocabulary mapping the OOV tokens to the original tokens in the sequence. This change will help me to both compute the loss function and the predictions generated by the model. Do you have any suggestions?

Thank you in advance for your reply!

Alessandro

**Comment by matt-gardner** [comment]
This is a good question, and I had to think a bit to come up with some ideas for how to do this.  Here's my initial take; there are probably ways to improve it.

As you say, there are two places where a copy mechanism needs special handling: in computing the loss function (should I have copied this word?) and when showing the output of decoding to a person.  These happen in very different places in the code, and should be handled separately.

First, for decoding, we typically [already pass through some metadata](https://github.com/allenai/allennlp/blob/a61aa67e363aef99bd4ae43af92bc8a2b3378de2/allennlp/models/biaffine_dependency_parser.py#L282) about what tokens were in the input, so that it can be displayed nicely in a demo.  You should be able to do the same thing during decoding in a copy model - you just keep track of whether you sampled from the vocab or from the copy mechanism during decoding, and display the result to the user accordingly, using that metadata field.  Doing it this way, instead of fudging with the vocabulary or input data, means that you can still use an arbitrary encoder to do your input embedding (including, e.g., ELMo), without resorting to any crazy hackery.

Second, the harder part, is how you train a model with a copy mechanism.  I _think_ the right thing to do here is to add a new `Field` that tells you token overlap on the target outputs.  For example, say I have the input sentence "the dog ate the food ." and the target sentence "the cat ate the mouse .".  What I need during training is to know which target tokens overlap with which input tokens, so my loss function can include the correct copy probabilities.  This field would output a tensor like this for those two inputs:

```
[[0, 3],    # the, appears twice in input
 [-1, -1],  # cat, not in the input
 [2, -1],   # ate, appears once in input
 [0, 3],    # the, appears twice in input
 [-1, -1],  # mouse, not in the input
 [5, -1]]   # ., appears once in input
```

where `-1` is used for padding here.

Then, in my loss computation, I'd do something like this:

```python
copy_log_probs = some_attention_softmax(on_some_inputs)
vocab_log_probs = some_softmax_over_vocab(based_on_hidden_state)
vocab_nll = cross_entropy(target, vocab_log_probs)
copy_targets_log_probs = copy_log_probs[copy_targets]  # I'm hand-waving at the indexing here
copy_nll = logsumexp(copy_targets_log_probs)
loss = logsumexp(mixing_prob.log() + vocab_nll, (1 - mixing_prob).log() + copy_nll)
```

Does this make sense?  Any thoughts?  @DeNeutoy, what do you think?

**Comment by matt-gardner** [comment]
Also, we would _love_ to have a nice implementation of a seq2seq + copy model in AllenNLP.  If you get this working, please consider contributing back.

**Comment by aleSuglia** [comment]
@matt-gardner: thanks for your feedback. At the moment I was checking the official implementation of See et al. (https://github.com/abisee/pointer-generator) and the copy mechanism implementation in OpenMT-py (https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/copy_generator.py). Apparently over there the idea is to treat the attention scores as a sort of probability score associated to the UNK tokens. Ultimately, they minimise the negative log-likelihood associated to the token indexes that compose the sequence plus additional special tokens associated to the unknowns tokens in the sentence.  

I'm going to work on this from now on. I'll let you know when I'll have a basic implementation. Are you thinking to integrate it in the current seq2seq implementation?

**Comment by matt-gardner** [comment]
Abigail's implementation makes one larger probability distribution from the vocab distribution and the attention distribution, seen here: https://github.com/abisee/pointer-generator/blob/a7317f573d01b944c31a76bde7218bcfc890ef6a/model.py#L146-L183

This accomplishes the same math as what I suggested, but requires a single distribution of size `vocab + num_OOVs` for each timestep.  Actually constructing this distribution is unnecessary, as you really just need to compute the loss.

The part that will be very tricky to make work with our data code is this line: https://github.com/abisee/pointer-generator/blob/a7317f573d01b944c31a76bde7218bcfc890ef6a/model.py#L172

Abigail's code has a separate input that includes token ids for the inputs with no OOVs computed.  You can do that with our data code - just have a separate `SequenceLabelField` for the input that uses the same vocabulary namespace as your target `SequenceLabelField`.  The trouble is we compute the vocabulary once, so if you ever want to run this on data that wasn't seen before (like in a demo), this could very easily break, because the tokens you want to copy aren't in the vocab.

So, these are the two options, I think:

1. Use my first suggestion of making a new `TokenOverlapField`, don't try to merge the two distributions, and just have the pointer loss be based on the overlap, without matching back to the vocabulary.
2. Have a separate `SequenceLabelField` for what Abigail calls the "extended vocab", and base this on a separate vocabulary namespace that's shared with your targets.  You might also need to have a separate `SequenceLabelField` for the extended vocab targets, actually, so that you can have a softmax just over the targets that are in your (non-extended) vocabulary...  Or you tie the input embeddings to the output embeddings, and solve the problem that way, though you'd have to make sure that the vocabularies match exactly between the inputs and the targets...

It feels like there are a lot more details that are easy to get wrong in the second way, especially as you think about putting up a demo for the model or using a trained model on new test data.

As far as integrating something with our current seq2seq implementation - that implementation was just an example, seeing if we could get any kind of seq2seq model implemented before starting some much more complex semantic parsing stuff.  I wouldn't suggest going out of your way to merge what you have with that, if you are starting from something else.

**Comment by matt-gardner** [comment]
@aleSuglia, @epwalsh just mentioned in #1928 that he has a copynet implementation available here: https://github.com/epwalsh/nlp-models/blob/master/modules/models/copynet.py.  He says it's only almost implemented, and I don't want to overload him with a support burden he didn't ask for, but that code may be a good starting place for you.

**Comment by aleSuglia** [comment]
@matt-gardner: thank you for mentioning it. I'll have a look asap :) 

I think we can close this issue!
